presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -240,6 +240,14 @@ public CostProvider getCostProvider()
             {
                 return costProvider;
             }
+
+            @Override
+            public void checkTimeoutNotExhausted()
+            {
+                if (((System.nanoTime() - context.startTimeInNanos) / 1_000_000) >= context.timeoutInMilliseconds) {
+                    throw new PrestoException(OPTIMIZER_TIMEOUT, format("The optimizer exhausted the time limit of %d ms", context.timeoutInMilliseconds));
+                }
```

Comment:
This logic duplicates the following logic in `IterativeOptimizer#exploreNode`. To avoid copy-paste, let's move `checkTimeoutNotExhausted` method to `IterativeOptimizer.Context#Context`.

```
            if (isTimeLimitExhausted(context)) {
                throw new PrestoException(OPTIMIZER_TIMEOUT, format("The optimizer exhausted the time limit of %d ms", context.timeoutInMilliseconds));
            }
```